libexec/apply_catalog.rb

```diff
@@ -67,9 +67,14 @@
            end
 
   Puppet.override(current_environment: env, loaders: Puppet::Pops::Loaders.new(env)) do
-    catalog = Puppet::Resource::Catalog.from_data_hash(args['catalog']).to_ral
+    catalog = Puppet::Resource::Catalog.from_data_hash(args['catalog'])
     catalog.environment = env.name.to_s
     catalog.environment_instance = env
+    if defined?(Puppet::Pops::Evaluator::DeferredResolver)
+      # Only available in Puppet 6
+      Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(Facter.to_hash, catalog)
```

Comment:
This may not be the best way to collect facts. We should probably use the facts terminus.